src/main/java/com/es/phoneshop/model/product/dao/impl/ArrayListProductDao.java

```diff
@@ -0,0 +1,95 @@
+package com.es.phoneshop.model.product.dao.impl;
+
+import com.es.phoneshop.model.product.dao.ProductDao;
+import com.es.phoneshop.model.product.entity.Product;
+import com.es.phoneshop.model.product.exception.ProductNotFoundException;
+
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.Currency;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.stream.Collectors;
+
+public class ArrayListProductDao implements ProductDao {
+    private List<Product> products;
+    private ReadWriteLock lock = new ReentrantReadWriteLock();
+
+    private long currentMaxId = 1;
+
+    public ArrayListProductDao() {
+        products = new ArrayList<>();
+        setSampleProducts();
+    }
+
+    @Override
+    public Product getProduct(Long id) throws ProductNotFoundException {
+        if (id == null)
+            throw new ProductNotFoundException("Parameter id is null");
+        lock.readLock().lock();
+        Product product = products.stream().
+                filter(currProduct -> id.equals(currProduct.getId())).
+                findAny().
+                orElseThrow(ProductNotFoundException::new);
+        lock.readLock().unlock();
+        return product;
+    }
+
+    @Override
+    public synchronized List<Product> findProducts() {
+        if (products.isEmpty())
+            return products;
+        return products.stream().
+                filter(product -> product.getPrice() != null).
+                filter(product -> product.getStock() > 0).
+                collect(Collectors.toList());
+    }
+
+    @Override
+    public void save(Product product) {
+        lock.writeLock().lock();
+        if (product.getId() != null) {
+            Optional<Product> productToAdd = products.stream().
+                    filter(product1 -> product1.getId().equals(product.getId())).
+                    findAny();
+            if (productToAdd.isEmpty()) {
+                product.setId(currentMaxId++);
```

Comment:
This is two lines of code repeated later in this method. You can extract them in additional method